src/dwn/models/record.js

```diff
@@ -0,0 +1,244 @@
+  get dataSize() { return this.#descriptor?.dataSize; }
+  get dateModified() { return this.#descriptor?.dateModified; }
+  get datePublished() { return this.#descriptor?.datePublished; }
+  get published() { return this.#descriptor?.published; }
+
+  /**
+   * Data handling.
+   */
+
+  get data() {
+    if (!this.#encodedData && !this.#readableStream) {
+      // `encodedData` will be set if `dataSize` <= DwnConstant.maxDataSizeAllowedToBeEncoded. (10KB as of April 2023)
+      // `readableStream` will be set if Record was instantiated from a RecordsRead reply.
+      // If neither of the above are true, then the record must be fetched from the DWN.
+      this.#readableStream = this.#dwn.records.read(this.#target, { author: this.#author, message: { recordId: this.#recordId } })
+        .then((response) => response.record )
+        .then((record) => { return record.data; });
+    }
+
+    if (typeof this.#encodedData === 'string') {
+      // If `encodedData` is set, then it is expected that:
+      // `dataSize` <= DwnConstant.maxDataSizeAllowedToBeEncoded (10KB as of April 2023)
+      // type is Uint8Array bytes if the Record object was instantiated from a RecordsWrite response
+      // type is Base64 URL encoded string if the Record object was instantiated from a RecordsQuery response
+      // If it is a string, we need to Base64 URL decode to bytes
+      this.#encodedData = Encoder.base64UrlToBytes(this.#encodedData);
+    }
+
+    const self = this; // Capture the context of the `Record` instance.
+    const dataObj = {
+      async json() {
+        if (self.#encodedData) return this.text().then(JSON.parse);
+        if (self.#readableStream) return this.text().then(JSON.parse);
+        return null;
+      },
+      async text() {
+        if (self.#encodedData) return Encoder.bytesToString(self.#encodedData);
+        if (self.#readableStream) return self.#readableStream.then(DataStream.toBytes).then(Encoder.bytesToString);
+        return null;
+      },
+      async stream() {
+        if (self.#encodedData) return DataStream.fromBytes(self.#encodedData);
+        if (self.#readableStream) return self.#readableStream;
+        return null;
+      },
+      then (...callbacks) {
+        return this.stream().then(...callbacks);
+      },
+      catch(callback) {
+        return dataObj.then().catch(callback);
+      },
+    };
+    return dataObj;
+  }
+
+  /**
+   * Record mutation methods.
+   */
+
+  async delete() {
+    if (this.isFrozen) throw new Error(`Error: Record with ID '${this.id}' was previously deleted.`);
+
+    // Attempt to delete the record from the DWN.
+    const response = await this.#dwn.records.delete(this.#target, { author: this.#author, message: { recordId: this.#recordId } });
+    
+    if (response.status.code === 202) {
+      // If the record was successfully deleted, freeze the instance to prevent further modifications.
+      this.#freezeRecord();
```

Comment:
i would suggest `#markAsDeleted` (and `#isDeleted`) as that's more clear as to what it does

Reply:
That is a better description.  Originally, the intent was to actually freeze the record in JavaScript.  However, the ability to undelete made that an issue since you cannot unfreeze a frozen record in javascript.

Addressed in d0a95c9f4d5b0c06bcaddea27921b750f9641701